pkg/csi_driver/multishare_ops_manager.go

```diff
@@ -280,327 +225,300 @@ func (m *MultishareOpsManager) startShareWorkflow(ctx context.Context, instanceS
+			shares, err := m.cloud.File.ListShares(ctx, &file.ListFilter{Project: instance.Project, Location: instance.Location, InstanceName: instance.Name})
+			if err != nil {
+				klog.Errorf("Failed to list shares of instance %s/%s/%s", instance.Project, instance.Location, instance.Name)
+				continue
+			}
+			if len(shares) >= util.MaxSharesPerInstance {
+				continue
+			}
+
+			readyEligibleInstances = append(readyEligibleInstances, instance)
+			klog.V(5).Infof("Adding instance %s to eligible list", instance.String())
 			continue
 		}
 
-		_, _, _, shareName, err := util.ParseShareHandle(string(item.Key))
-		if err != nil {
-			return false, err
+		// instances whose delete is in progress should not be accounted as non-ready instances.
+		if op.Type == util.InstanceDelete {
+			continue
 		}
 
-		opInfo, status, err := m.checkAndUpdateShareOp(ctx, instanceScPrefix, &file.Share{Parent: instance, Name: shareName})
-		if err != nil {
-			return false, err
-		}
-		if opInfo != nil && status == util.StatusRunning {
-			return false, nil
+		if instance.State == "DELETING" {
+			continue
 		}
+		nonReadyInstanceCount += 1
+		// TODO: If we see > 1 instances with 0 shares (these could be possibly leaked instances where the driver hit timeout during creation op was in progress), should we trigger delete op for such instances? Possibly yes. Given that instance create/delete and share create/delete is serialized, maybe yes.
 	}
 
-	return true, nil
+	return readyEligibleInstances, nonReadyInstanceCount, nil
 }
 
-func (m *MultishareOpsManager) validateShareExists(ctx context.Context, instanceHandle, shareName string) (*file.Share, error) {
-	project, location, instanceName, err := util.ParseInstanceHandle(instanceHandle)
+func (m *MultishareOpsManager) instanceNeedsExpand(ctx context.Context, share *file.Share) (bool, int64, error) {
+	if share == nil {
+		return false, 0, fmt.Errorf("empty share")
+	}
+	if share.Parent == nil {
+		return false, 0, fmt.Errorf("Parent missing from share %q", share.Name)
+	}
+
+	shares, err := m.cloud.File.ListShares(ctx, &file.ListFilter{Project: share.Parent.Project, Location: share.Parent.Location, InstanceName: share.Parent.Name})
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "failed to parse instance handle %v", instanceHandle)
+		return false, 0, err
 	}
 
-	return m.cloud.File.GetShare(ctx, &file.Share{
-		Name: shareName,
-		Parent: &file.MultishareInstance{
-			Project:  project,
-			Location: location,
-			Name:     instanceName,
-		},
-	})
+	var sumShareBytes int64
+	for _, s := range shares {
+		sumShareBytes = sumShareBytes + s.CapacityBytes
+	}
+	// TODO: Check if we need to align the increment to step size.
+	var remainingBytes int64
+	remainingBytes = share.Parent.CapacityBytes - sumShareBytes
+	if remainingBytes < share.CapacityBytes {
+		return true, share.Parent.CapacityBytes + (share.CapacityBytes - remainingBytes), nil
```

Comment:
Here we can just use `sumShareBytes + share.CapacityBytes` as the new target cap because if you substitute `ramainingBytes` with how it's calculated, `share.Parent.CapacityBytes` will cancel out

Reply:
done